### ISSUE TITLE
Fade transform

### ISSUE THREAD

**Issue by Quefumas** [Quefumas/gensound#8]
Each of these can be dealt with separately, do not fear:

1. How are fade curves typically implemented? exponential, linear change? in amplitude or dB? If the curve is by dB, how do we reach -inf? IIRC Reaper provides a good selection of fades, we can try to aim for these options.

2. How should the fade curve be chosen by the user? Do we give a string as an argument, i.e. `Fade(curve='linear')`? Which one should be the default? Can we have the user replace the default manually?

3. (*Perhaps should be split into a separate issue in the future*) I'm not entirely pleased with the `is_in` argument, which indicates if the user desires fade in or out. I believe we can find a better solution syntactically. One option coming to mind is to simply split it into 2 transforms, `FadeIn` and `FadeOut`, and that could be the default solution for now. A more tempting idea is to have the fade direction determined by the order of multiplication with the signal. For example: `Fade()*sig` is interpreted to mean fade in, while `sig*Fade()` is taken to mean fade out, implicitly.
This solution is very nice for the user, but below the scenes it requires a significant break of the syntax rules, since the Transform has to be somehow notified of the side on which it was applied to the signal. Even if we supply each signal with 2 transform list, pre- and post- (as opposed to a single list like right now), still the signal will have to notify the transforms when realising.
Do note however that there is somewhat of an exception to this multiplication rule, since we can left-multiply by a float, while other Transforms have to be right-multiplied. Still, this exception is allowed not only because amplitude change is an essential transform, but also because it admits this syntax which simply hides its being a transform, so it is more acceptable.

**Comment by Quefumas** [comment]
Syntax suggestion (for 3): How about like this:
```python
s = sig*Fade()[:0.5e3]
```
I.e. using slice notation to indicate fade duration? In that case for fade out we can use:
```python
s = sig*Fade()[-0.5e3:]
```

Another thought: there isn't a very good reason why Transforms shouldn't be multipliable on both sides of a Signal, so maybe we should just allow it? and maintain the convention that only amplitude and Fadeins should be placed on the left?

**Comment by zivkaplan** [comment]
I believe it is right to split the`Fade()` into two separate`FadeIn()`, `FadeOUt()`.
I imagine the following behavior: 
Each fade will receive a two parameters, a `duration` parameter (currently defaults to 3e3 if not specified by the user) and `curve` parameter (I believe now it is defaults to linear). The curve parameter will be able to take one of three values: `Linear`, `Exponential` or a custom curve that will be passed as a python dictionary. The dictionary will contain keys as time markers and their values as the desired curve (linear/ exponential) from the specified time maker as the key to the next.
The keys, meaning the time markers, will be integers between 0 - 100, marking the desired place on percentages of the specified duration (similar to a CSS animation).

example for a `FadeIn` with a custom curve as described above:
`wav *= FadeIn(duration=5000, curve={0: "linear", 70: 'exponential', })`

meaning, the fade will be linear from the start (0%) up to 70% of the specified duration, then it will be exponential till the end of the fade duration.

**Comment by zivkaplan** [comment]
Regarding the way of conveying a custom fade curve:

In the message above the dictionary was read as:
" **from the (specified key) position:**  do an (specified value)curve".
Thus, the user never use the 100% marker, and will always have to mark a 0% in order to let the fade know the first wanted curve. 
If it is better, the logic can be flipped to:
"**up to (specified key) position**: do an (specified value) curve". 
Thus, the user will always have to use the 100% marker to cover to whole fade duration, but he would not have to specify a 0%.

For example,
a curve of linear fade to the 50% position and exponential till the 100% will look like this:
Using the first logic:
`{0: "linear", 50: "exponential",}`
Using the second logic:
`{50: "linear", 100: "exponential", }`

I do believe the **first logic is better**, since that way, the program can have a default linear curve for the if not specified a 0% mark, making a possible short input from the user. 
for example, the fade written above could be shortened to: 
`{50: "exponential",}`
from the 50% place do an exponential curve. before the 50% will have a defaulted linear curve since it was not specified as different from the user.

**Comment by Quefumas** [comment]
Hello and welcome! I appreciate hearing your thoughts about this issue.

As for no. 3 above (separating `Fade` into `FadeIn` and `FadeOut`), I think this is probably the best way, as it is the most explicit, isn't too wordy, and doesn't break anything. Also it does not necessarily invalidate a future feature for inferring the direction of the fade by syntax alone, as proposed. Feel free to go ahead and suggest an implementation 👍.


As for specifying the curve, I'm not entirely on board with this suggestion in its current form. To be more specific:
1. Why are the dictionary keys percentages, and not absolute time indications? It's not immediately obvious to the user, and it is conceivable that users would want to specify actual durations (in ms or samples) instead of percentages, leading to confusing behaviour. Also, what happens if the user supplies a key greater than 100?
2. Are we sure this syntax is flexible enough? [Reaper](https://www.reaper.fm) offers no less than 7 (!) fade curves, and I think it's likely that many of the possible curves have additional parameters to adjust steepness (there is more than one possible way to implement an exponential curve). In that case, it's not clear at all how this dictionary would accommodate these extra parameters.
3. How exactly are we supposed to combine the different kinds of curves? For example, your example of 50-50 division between linear and exponential curves does not make it obvious what would be the amplitude level at the 50% mark - would the linear part bring us from 0 to 0.5, and the exponential part do the rest? or will the linear part grow slowly from 0 to 0.1? This is especially confusing for the non-linear curves, as we can use them from different starting points on the curve.
4. Finally, there are already tools that can achieve essentially the same effect:
```python
from gensound.transforms import Amplitude
from gensound.curve import Line, Logistic

s = WAV(test_wav)[15e3:25e3]
c = Line(0.0, 0.1, duration=2e3) | Logistic(0.1, 1, duration=2e3)
s *= Amplitude(c)
s.play(sample_rate=24000)
```
Admittedly, this isn't documented anywhere yet (and it should be).
This may seem somewhat complex to use, but I imagine that this feature would be relevant to advanced users, which would appreciate the high level of control. Note that this syntax is compatible with points 1-3, and is also compatible with normal Signal concatenation, while the dictionary format is not used anywhere else.


On an aside, I would prefer your "first logic", but mainly for visual reasons: it is easy to see that there is a linear curve between 0 to 50, while in the 2nd logic this can give the impression that there is a linear curve between 50 and 100, and thus requires more brain power to understand. There is something to be said for the 2nd logic as well though, since using this syntax it is always clear to the user that they are using percentages (since the key 100 must always be present), which clears some of the confusion. But again see point 1 above.

**Comment by zivkaplan** [comment]
Hi! 
1. glad to hear that we agree on the `FadeOut` \ `FadeIn` splitting!

2. Percentages \ absolute time indications:
What is the main use that we are expecting the user to use this package and specific method? 
If the user is trying to have a millisecond control, will he use a python? wouldn't he use any DAW? 
to sum my point: 
- If you designate the package to be a python audio editing for the average user (simple editing, prototyping, mass editing?) then percentages is better in my opinion. 
Percentages do require the user to read the docs (as he should anyhow), but, it is a simple concept to explain and an easy one to understand and implement (from the users perspective). It is much more friendly with percentages with a decent amount of control. 
- If you designate it to be a package that is used by professionals, that will use it as a foundations to build an actual DAW, with a different user interface, then I agree an absolute time is better. 
 
anyway, it absolute time markers are crucial to have, can't we implement both? maybe specifying a parameter for choosing the preferred method, that will be default to one of those that you believe is better.

3. regarding 100% and up. 
I believe it is a solvable issue, and is one that is exclusive to percentages. a user can define a `duration` parameter that will conflict with his custom `curve` inputs also in absolute time markers.
I see two solutions here (for percentages):
- validating the keys and raising an error. 
- Auto fixing (in our `curve`  syntax we specific a starting point marker and a curve style, so the last valid input will last automatically last up to 100%, ignoring all invalid keys.) and maybe printing a warning? 

4. I have now experience with Reaper and I don't know about its 7 (!) fade curves! DO you know any resources that describe these? 
maybe we can replicate them using various combinations of linear and exponential curves and have some presets for our users. 

5. what effect exactly is achieved in your example in 4? 

I hope that my questions will lead us to the right directions!

**Comment by Quefumas** [comment]
I'll start from the end, and glide into the other points on the way.
The code above applies fade in by directly multiplying the signal amplitudes by a curve which grows over time (just like any fade in). Specifically, the amplitude starts at zero and grows to 0.1 of the original value within 2 seconds, and then takes another 2 seconds to grow from 0.1 to the full signal amplitude, but this time following a [logistic curve](https://en.wikipedia.org/wiki/Logistic_function).
In other words, this already exists and has the same effect as the suggested feature, and with more direct control over all parameters. The only apparent disadvantage is not being able to specify the change in percentages, but only in absolute time, but in that case it would be still quite simple to do the following:
```python
from gensound.transforms import Amplitude
from gensound.curve import Line, Logistic

s = WAV(test_wav)[15e3:25e3]
fade_duration = 3e3
c = Line(0.0, 0.1, duration=0.7*fade_duration) | Logistic(0.1, 1, duration=0.3*fade_duration)
s *= Amplitude(c)
s.play(sample_rate=24000)
````
Now we have 70%-30% between the two curves, and we only need to change one variable to adjust the size of 100%.
Maybe this looks a bit too much work, but I think that a user that actually wants to go into these little details should be prepared to deal with it.

As for use-cases and DAWs and everything - keep in mind that the audio is always generated offline, never in real time, and plus Python isn't really fast enough, so this library is absolutely not designed to build a professional DAW from. We already have great DAWs, and Reaper is even programmable up to a point. But more importantly, exactly the kind of feature we're discussing right now is what Gensound makes possible and easy, but would be quite inconvenient in any actual DAW, because it brings about a lot of GUI design issues. Gensound's target audience is programmers who make music, not music producers, so we allow ourselves to use all of the advantages of typing stuff out instead of having a GUI. As for "audio editing for the average user" - I somewhat doubt that a casual user would ever need to fine-tune up to the point of manually cutting up and connecting different types of curve together.

To summarize, I think that this more advanced feature is a design idea we can keep in mind and may be useful to us at some point in the future, maybe here or maybe for a different feature, but we shouldn't rush to implement it. The main reason is that there is already a way to do this exact same thing, which already offers ways to deal with all of the issues raised above, and doesn't appear to be clearly inferior to the suggested syntax.

Either way, I suggest that we first focus on having a solid implementation for the "usual" fades, that is, a curve of just a single type. There is work to do here - I don't think we should be in any rush to offer the 7 options available in Reaper (which I'm also not exactly sure how they are defined, perhaps I can guess from the graphic), but it would be really useful to know what the "classical"/common practice options are and implement those for a start. Reaper's source code isn't available for examination, but Audacity is open source, so that is one place we can look.

**Comment by zivkaplan** [comment]
Now I understand! Great example. 

A question: 
eventually, in your example, the fade was not a part of the full audio file. 
you declared a separated variable containing only the beginning of the audio, and applying a fade on that.
Is there a way of doing a fade the same way you showed directly on the full audio variable? (as the `Fade` is applied? )

Assuming it is possible - 
If so, then the `Fade` isn't more then just a 'syntactic sugar'! We are after is actually a shortcut for something that it already implement in Gensound, however might be a bit complicated and currently isn't well documented.
Knowing that, should we implement a `fade` at all? What will be its advantage over using a "spelled  out" code as you example above, assuming it will be well documented with examples for several scenarios? 
The user who will have a strong opinion about his effects can learn how to implement a desired custom fade himself. 
Maybe, the `fade` function should be framed as 'Editing Presets' part ot the Gensound package? for the less demanding user who does not care for specific curve and looking for a simple and quick solutions? Maybe with other tools that work "right out of the box"?

In that scenraio, I believe we should limit the functions capabilities on purpose, letting its `curve` parameter accept only non-custom options such as "linear" and exponential" (or more,regarding the Reaper/Audacity research results). (`duration` parameter stays the same of course).

**Comment by Quefumas** [comment]
> Is there a way of doing a fade the same way you showed directly on the full audio variable? (as the Fade is applied? )

You can use ```s = WAV(test_wav)``` just as well, if that's what you mean. I sliced away the introduction to make the Fade more apparent.

>  Then the fade we are after ... a nicer way for something that it already implement in Gensound ... a bit complicated and currently isn't well documented.

Yes, or rather: what I showed above is the overly complicated and usually unnecessary way to do Fades, which are already implemented.

>  What will be its advantage...? ... for the less demanding user who does not care for specific curve and looking for a simple and quick solutions? Maybe with other tools that work "right out of the box"? ... The user who will have a strong opinion about his effects can learn how to implement a desired custom fade himself.

You answered it yourself!

>  I believe we should limit the functions capabilities on purpose

Yes. We can enable more complex curves if: 1. we find a very elegant way to offer this option, and 2. they are likely to be used. Both appear to be false, and additionally there already exists a way to do this.

The thing is, we absolutely can't NOT provide a simple Fade capability, since this is used extremely often, and any user with some knowledge of audio would expect it to exist.
It's true that using ```gensound.curve``` seems to cover that, but that doesn't bother me, since it can be used for so much more than just fades, and it is there only for the users who would know to look for it (once I get it documented, that is).

**Comment by zivkaplan** [comment]
Good! 
So our conclusions are:
1. Improve documentation to include sufficient explanations about a user-made custom fade using the transforms and curves libraries. 
2. Build a limited `FadeIn` and `FadeOut`for the average user for "day to day" use. steps in the way:
-Decide on the functions scale - it's options and limits. 
I suggest to start building by duration and curve (linear of exponential) with no other custom options. Will the user who will want a built-in option needs more? I believe it is good enough for the majority at least. 

What do you think?

**Comment by Quefumas** [comment]
Agreed.

No. 1 is on me, but it will have to wait, since the curves are actually a large and advanced set of features that only incidentally can be used to implement unorthodox fades. I feel the curves, although quite functional, still require more consideration before official release. It's likely I will first provide more basic documentation along with backward compatibility warnings. For the issue at hand it is not urgent to document the curves, but the classical Fade should be documented as soon as it's ready (along with other essential Transforms which are still missing from the wiki).

As for 2, feel free to go ahead with a suggested implementation! I agree that for a start (or maybe for ever) it should only allow a choice of a single curve type for the entire fade (i.e. no mixing between linear and exponential).

Some things to consider for the implementation:
1. Should the curve argument be of style `curve="linear"`, or rather `curve=gensound.LINEAR_CURVE`? I think most people (me included) would prefer the first, but do note that the 2nd syntax is quite popular in some languages, as it avoids strings (maybe because then a typing error will raise an exception instead of fail silently?)
2. I agree there's no need to support more than 2-3 curve types right now, but the current choices of `linear` and `exponential` as well as their mathematical computation are no more than educated guesses. It would be best if we can figure out what people use most often, and how it translates mathematically (to amplitude or dBs). If we end up including two options which nobody uses in real life, that would be a shame and may result in breaking backward compatibility when we come back to fix it.
3. It is quite possible that at least the exponential curve is tweakable, even in common usage (e.g. we can use both `x**2` as well as `x**3` to implement fades, they are both exponential in some sense of the word). So even if we don't give the user an option to further characterize the curve, we should keep it in mind for the future. For example, having a `steepness` argument may be able to serve several possible curves.

I think the main issue is point no. 2 above, once we have that figured out, it will be easy to know what to do about 3.

**Comment by zivkaplan** [comment]
Great, 
I too think passing a string is better. 
Maybe its a rookie way of thinking, but we can, at least at the beginning, when we should have only small number of valid inputs, check first if the string appears in a valid inputs array or something like that. If not, we can raise an exception of print an error).

regarding the fades themselves: 
For a start, here is [Audacity manual page for fades](https://manual.audacityteam.org/man/fades.html). 
I will look for more info, and try to better understand how the fade function currently works and how to implement our ideas.

**Comment by Quefumas** [comment]
[This](https://manual.audacityteam.org/man/fade_and_crossfade.html) page from the manual seems to show a comprehensive list of the usable curves, and can give us a better idea of what users would expect to have available. Unfortunately it doesn't give the maths precisely. [This one](https://manual.audacityteam.org/man/adjustable_fade.html) shows what more advanced users may expect to have available, and it seems that some of the implementation code is [here](https://github.com/audacity/audacity/blob/95a7fc6362729594565204b6db927b2e025d20c0/plug-ins/adjustable-fade.ny), but in a Lisp-like scripting language.

For the near future I think we can deal quite well with just having linear, exponential and "equal power" fade. The latter will be very useful because it may be necessary for the `CrossFade` Transform, which already exists but doesn't work well (because of this issue exactly).

All the S-curves, that is, sines, cosines, logistic functions, tanh, whatever, these can all wait, but it would be good to mess around with Audacity and see what options are given for tweaking them in order to see if we can get a uniform interface of arguments in preparation for all kinds of curves. In particular, they mention applying linear fade several times, I think we may want to support this option.

**Comment by zivkaplan** [comment]
Question: 
I am yet to know the ins and outs of Gensound. 
Does it has an exponential curve in the transforms or we need to write it?

**Comment by Quefumas** [comment]
Some ideas about possible curves:
* `curve='linear'` or `'lin'` would mean linear amplitude growth, that is, multiplying the amplitudes by `x` as `x` grows from 0 to 1.
* `curve='polynomial'` or `'poly'` will give the effect of applying a linear fade multiple times, which would be equivalent to multiplying amplitudes by `x**a` for some `a>0`, as `x` grows from 0 to 1. We can then provide a `steepness` or `degree` argument, so the user can choose the value of `a`. Setting `a=1` will naturally result in the linear curve. This parameter can also be a float, which allows for fine-tuning.

I'm not exactly sure yet what definitions they use for 'exponential', 'log-exp' etc. curves, but these options may be relevant, don't know how they are called:
* `amp(x) = (e**x - 1) / (e - 1)`, this is very similar to `x**1.5` in the range 0 to 1.
* Make the dBs grow linearly rather than the amplitudes, which means taking equal steps between a somewhat arbitrary starting point (for example -60 dB), and gradually raising it to 0, then converting this range to the appropriate amplitudes. This has the disadvantage that the 'silence' starting point is quite arbitrary, as absolute silence is minus infinity in dBs.


Here are also two questions from StackExchange that are relevant: https://dsp.stackexchange.com/questions/14754/equal-power-crossfade, https://dsp.stackexchange.com/questions/37477/understanding-equal-power-crossfades.

**Comment by zivkaplan** [comment]
Hi,

I have a suggestion. Not sure it is better. 
Since as you said,  there are many kind of curves for fades the area is not so standardized, 
we can drop the usage on names and use numbers in the curve parameter value.
The down side is that it will be less explicit for simple curve as linear, 
but it will allow the function to be more scalable since
1. we are not forced to make up names for each curve style. 
2. We can use some helper function or dictionary to easy store and update many kinds of curves.

What do you think?

**Comment by Quefumas** [comment]
I'm not sure how using numbers would be any better - they're definitely less informative to the user, and we can validate strings just as well as ints. I agree that there will be a certain making-up of names, but they can still be made quite suggestive as to the actual curve. As for helper functions for storing curves, we can use a dict with strings as keys just as well, so I don't see a big difference either. As for documentation - we will need it anyway, but forcing the user to open it is something we usually want to avoid (we want it self-explanatory). I estimate that using ints will make the user spend more time digging and less composing.

Also, I am getting the feeling that we won't have much to gain from having a uniform "adjustment" argument for different kinds of curves; instead we should just allow each curve to have its own set of extra arguments, when needed.

**Comment by zivkaplan** [comment]
Question regarding your comments for my last pull request:

 the `polynomial` as I implemented it  - is the `linear` equation with ** at the end. 
  - Should I find a way to make the `polynomial` use a reference to the `linear` as a key, so it will not be written twice?
The problem is that I couldn't find a way to reference a dict inside itself.
  - Or: should we use only the `polynomial` and treat it a one type of fade with optional `degree` parameter for tweaking. `degree = 1` is `linear`. 

```
curve_types = {
            "linear": np.linspace(0, 1, self.num_samples(audio.sample_rate)),
            "polynomial" : (np.linspace(0, 1, self.num_samples(audio.sample_rate))) ** self.degree,
        }
```

**Comment by Quefumas** [comment]
I don't think a `dict` is the way to go here. For one, saving a couple of `if` lines doesn't cover the cost of having code which is harder to read. Short code should be for the sake of clarity, not for its own sake.
And talking about efficiency, this code means evaluating each and every one of the available curves, which is very wasteful (imagine when there are 10+ curve types and someone tries to perform gradual fade over 10 minutes of audio - this means performing all sorts of algebra on 10 series of length over a million, only to throw away most of them). Also consider that if some arguments like degree are missing, this may cause runtime issues.

As for `polynomial` with degree 1, I say just let it be. The code will be clearer, and besides, I think we can trust NumPy to efficiently evaluate `x**1`.

**Comment by zivkaplan** [comment]
1. thanks for your input about efficiency, I didn't take that into account. 

2. I left the curves as two separated equations.
```
 if self.curve == "linear": amp = np.linspace(0, 1, self.num_samples(audio.sample_rate))
 if self.curve == "polynomial": amp = (np.linspace(0, 1, self.num_samples(audio.sample_rate))) ** self.degree      
```
I
f the code was: 
``` 
if self.curve == "linear" or "polynomial":
           (np.linspace(0, 1, self.num_samples(audio.sample_rate))) ** self.degree
```
There is no validation that if `curve="linear"` then `degree = 1`. `degree` is indeed  initialized to`1`, but a user can still specify `curve=linear degree=0.5`. 
It won't break it of course, but it doesn't make sense to me giving that option to the user - it is eliminating the need for a curve type at all (at least for now, that we have only 2 with similar enough equation).

see update pull request! 

another question: should we allow `degree=0`?

**Comment by Quefumas** [comment]
If the curve is linear, we do not actually care about the value of degree - it should be ignored, and there's actually not much we can do about it. We can consider raising a warning if the user decided to specify some irrelevant parameters. (Also see comment relating to this on the pull request)

As for degree 0, I haven't made up my mind yet. What is your opinion? (I assume you realise how the fade would sound like if the degree is 0!)

**Comment by Quefumas** [comment]
I wrote some comments on the code, but I have some more I'll add in a couple of hours.
